File: src/main/java/mil/nga/tiff/internal/TileOrStripProcessor.java
package mil.nga.tiff.internal;

import mil.nga.tiff.compression.CompressionDecoder;
import mil.nga.tiff.field.type.enumeration.Compression;
import mil.nga.tiff.field.type.enumeration.PlanarConfiguration;
import mil.nga.tiff.io.ByteReader;

import java.nio.ByteOrder;

public class TileOrStripProcessor {

    private final DirectoryStats stats;
    private final TileOrStripCache cache;


    /**
     * Create tile or strip processor
     *
     * @param stats directory stats
     * @param cache true to cache tiles and strips
     */
    public TileOrStripProcessor(DirectoryStats stats, TileOrStripCache cache) {
        this.stats = stats;
        this.cache = cache;
    }

    /**
     * Get the tile or strip for the sample coordinate
     *
     * @param x      x coordinate
     * @param y      y coordinate
     * @param sample sample index
     * @return bytes
     */
    public byte[] run(int x, int y, int sample, ByteReader reader, boolean tiled, ByteOrder byteOrder) {
        int index = determineIndex(x, y, sample);

        return cache.getOrSet(index, () -> {
            // Read and decode the block
            long offset;
            int byteCount;
            byte[] tileOrStrip;

            if (tiled) {
                offset = stats.tileOffsets().get(index);
                byteCount = stats.tileByteCounts().get(index);
            } else {
                offset = stats.stripOffsets().get(index);
                byteCount = stats.stripByteCounts().get(index);
            }

            reader.setNextByte(offset);
            byte[] bytes = reader.readBytes(byteCount);
            CompressionDecoder decoder = Compression.getDecoder(stats.compression());
            tileOrStrip = decoder.decode(bytes, reader.getByteOrder());

            tileOrStrip = stats.predictor().getImplementation().decode(
                tileOrStrip,
                stats.tileWidth(),
                stats.tileHeight(),
                stats.bitsPerSample(),
                stats.planarConfiguration(),
                byteOrder
            );

            // Cache the data
            return tileOrStrip;
        });
    }

    private int determineIndex(int x, int y, int sample) {
        int imageWidth = stats.imageWidth();
        int imageHeight = stats.imageHeight();
        int tileWidth = stats.tileWidth();
        int tileHeight = stats.tileHeight();
        int numTilesPerRow = (imageWidth + tileWidth - 1) / tileWidth;
        int numTilesPerCol = (imageHeight + tileHeight - 1) / tileHeight;

        if (stats.planarConfiguration() == PlanarConfiguration.CHUNKY) {
            return y * numTilesPerRow + x;
        } else if (stats.planarConfiguration() == PlanarConfiguration.PLANAR) {
            return sample * numTilesPerRow * numTilesPerCol + y * numTilesPerRow + x;
        }

        return 0;
    }

}

File: src/main/java/mil/nga/tiff/internal/TileOrStripProcessor.kt
package mil.nga.tiff.internal

import mil.nga.tiff.field.type.enumeration.Compression
import mil.nga.tiff.field.type.enumeration.PlanarConfiguration
import mil.nga.tiff.io.ByteReader
import java.nio.ByteOrder

/**
 * Create tile or strip processor
 *
 * @param stats directory stats
 * @param cache true to cache tiles and strips
 */
class TileOrStripProcessor(private val stats: DirectoryStats, private val cache: TileOrStripCache) {

    /**
     * Get the tile or strip for the sample coordinate
     *
     * @param x      x coordinate
     * @param y      y coordinate
     * @param sample sample index
     * @return bytes
     */
    fun run(x: Int, y: Int, sample: Int, reader: ByteReader, tiled: Boolean, byteOrder: ByteOrder): ByteArray {
        val index = determineIndex(x, y, sample)

        return cache.getOrSet(index) {
            // Read and decode the block
            val offset: Long
            val byteCount: Int
            var tileOrStrip: ByteArray?

            if (tiled) {
                offset = stats.tileOffsets!![index]
                byteCount = stats.tileByteCounts!![index]
            } else {
                offset = stats.stripOffsets!![index]
                byteCount = stats.stripByteCounts!![index]
            }

            reader.setNextByte(offset)
            val bytes = reader.readBytes(byteCount)
            val decoder = stats.compression.decoder()
            tileOrStrip = decoder.decode(bytes, reader.byteOrder)

            tileOrStrip = stats.predictor.implementation.decode(
                tileOrStrip,
                stats.tileWidth!!,
                stats.tileHeight!!,
                stats.bitsPerSample!!,
                stats.planarConfiguration!!,
                byteOrder
            )
            tileOrStrip
        }
    }

    private fun determineIndex(x: Int, y: Int, sample: Int): Int {
        val imageWidth = stats.imageWidth!!
        val imageHeight = stats.imageHeight!!
        val tileWidth = stats.tileWidth!!
        val tileHeight = stats.tileHeight!!
        val numTilesPerRow = (imageWidth + tileWidth - 1) / tileWidth
        val numTilesPerCol = (imageHeight + tileHeight - 1) / tileHeight

        if (stats.planarConfiguration == PlanarConfiguration.CHUNKY) {
            return y * numTilesPerRow + x
        } else if (stats.planarConfiguration == PlanarConfiguration.PLANAR) {
            return sample * numTilesPerRow * numTilesPerCol + y * numTilesPerRow + x
        }

        return 0
    }
}
